man/comp.man
.TH COMP %manext1% 2022-12-22 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
comp \- compose an nmh message
.SH SYNOPSIS
.HP 5
.na
.B comp
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msg ]
.RB [ \-form
.IR formfile ]
.RB [ \-use " | " \-nouse ]
.RB [ \-file
.IR file ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-editor
.IR editor ]
.RB [ \-noedit ]
.RB [ \-width
.IR columns ]
.RB [ \-from
.IR address ]
.RB [ \-to
.IR address ]
.RB [ \-cc
.IR address ]
.RB [ \-fcc
.IR +folder ]
.RB [ \-subject
.IR text ]
.RB [ \-whatnowproc
.IR program ]
.RB [ \-nowhatnowproc ]
.RB [ \-build ]
.ad
.SH DESCRIPTION
.B comp
is used to create a new message to be mailed.  It copies a
message form to the draft being composed and then invokes an editor on
the draft (unless
.B \-noedit
is given, in which case the initial edit is suppressed).
.PP
The default message template \*(lqcomponents\*(rq will direct
.B comp
to construct the message draft as follows:
.PP
.RS 5
.nf
From: {from switch} or <Local-Mailbox> or <username@hostname>
To: {to switch} or blank
cc: {cc switch} or blank
Fcc: {fcc switch} or +outbox
Subject: {subject switch} or blank
--------
.fi
.RE
.PP
If a file named
.RI \*(lq components \*(rq
exists in the user's nmh directory,
it will be used instead of this form.  You may specify an alternate
forms file with the switch
.B \-form
.IR formfile .
.PP
You may also start
.B comp
using the contents of an existing message
as the form.  If you supply either a
.I +folder
or
.I msg
argument, that
message will be used as the message form.  You may not supply both a
.B \-form
.I formfile
and a
.I +folder
or
.I msg
argument.  The line of
dashes or a blank line must be left between the header and the body of
the message for the message to be identified properly when it is sent
(see
.IR send (1)).
.PP
The switch
.B \-use
directs
.B comp
to continue editing an already
started message.  That is, if a
.B comp
(or
.BR dist ,
.BR repl ,
or
.BR forw )
is terminated without sending the draft, the draft can
be edited again via
.RB \*(lq comp
.BR \-use \*(rq.
.PP
The
.B \-file
.I file
switch says to use the named file as the message draft.
.PP
Forms that are selected via the
.B \-form
switch are processed via the
.B nmh
template system; see
.IR mh\-format (5)
for details.  Drafts constructed from another message or with the
.B \-use
or
.B \-file
switches will
.I not
be processed with
.IR mh\-format (5).
.PP
In addition to the standard
.IR mh\-format (5)
escapes,
the following
.I component
escapes are either new or have an alternate meaning:
.PP
.RS 5
.nf
.ta \w'Escape  'u +\w'Returns  'u
.I "Escape	Returns	Description"
fcc	string	Any folders specified with `\-fcc\ folder'
from	string	Any addresses specified with `\-from\ address'
to	string	Any addresses specified with `\-to\ address'
cc	string	Any addresses specified with `\-cc\ address'
subject	string	Any text specified with `\-subject\ text'
.fi
.RE
.PP
By default the \*(lqTo:\*(rq and \*(lqcc:\*(rq fields are empty.  You may
add addresses to these fields with the
.B \-to
.I address
and
.B \-cc
.I address
switches.  You may give these switches multiple times to add multiple
addresses.
.PP
By default the \*(lqFrom:\*(rq field has either the value of the
.B Local\-Mailbox
profile entry or a system default email address.  This default can be
overridden by using the
.B \-from
.I address
switch.  The default mailbox in the \*(lqFcc:\*(rq field is
.IR +outbox .
This can be overridden by the
.B \-fcc
switch.
.PP
Any text you give to the
.B \-subject
switch will be placed in the \*(lqSubject:\*(rq field in the draft.
.PP
If the draft already exists,
.B comp
will ask you as to the disposition
of the draft.  A reply of
.B quit
will abort
.BR comp ,
leaving the draft intact;
.B replace
will replace the existing draft with
the appropriate form;
.B list
will display the draft;
.B use
will use the draft for further composition; and
.B refile
.I +folder
will file the draft in the given folder, and give you a new draft with the
appropriate form.  (The
.I +folder
argument to
.B refile
is required.)
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke the
.B nmh
draft folder facility.  This is an advanced (and highly
useful) feature.  Consult
.IR mh-draft (5)
for more information.
.PP
The
.B \-editor
.I editor
switch indicates the editor to use for the
initial edit.  Upon exiting from the editor,
.B comp
will invoke the
.B whatnow
program (see
.IR whatnow (1)
for a discussion of
available options).  The invocation of this program can be inhibited
by using the
.B \-nowhatnowproc
switch.  (In fact, it is the
.I whatnow
program which starts the initial edit.  Hence,
.B \-nowhatnowproc
will prevent any edit from occurring.)
.PP
The
.B \-build
switch is intended to be used by the Emacs mh-e interface to
.BR nmh .
It implies
.BR \-nowhatnowproc .
It causes a file <mh-dir>/draft to be created, containing the draft message that
would normally be presented to the user for editing.
No mail is actually sent.
.SH FILES
.PD 0
.TP 20
%nmhetcdir%/components
The default message skeleton.
.TP
<mh-dir>/components
The user's message skeleton.
.TP
$HOME/.mh_profile
The user's profile.
.PD
.SH "PROFILE COMPONENTS"
.PD 0
.TP 20
Path:
To determine the user's nmh directory.
.TP
Draft\-Folder:
To specify the default draftfolder.
.TP
Editor:
To override the default editor.
.TP
Msg\-Protect:
To set mode when creating a new message (draft).
.TP
fileproc:
Program to refile the message.
.TP
whatnowproc:
Program to ask the \*(lqWhat now?\*(rq questions
.PD
.SH "SEE ALSO"
.IR dist (1),
.IR forw (1),
.IR repl (1),
.IR send (1),
.IR whatnow (1),
.IR mh\-draft (5),
.IR mh\-profile (5)
.SH DEFAULTS
.PD 0
.TP 20
+folder
The current folder.
.TP
msg
The current message.
.TP
\-nodraftfolder
.TP
\-nouse
.PD
.SH CONTEXT
None
.SH BUGS
If
.I whatnowproc
is
.BR whatnow ,
then
.B comp
uses a built-in
.BR whatnow ,
it does not actually run the
.B whatnow
program.
Hence, if you define your own
.IR whatnowproc ,
don't call it
.B whatnow
since
.B comp
won't run it.


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/comp.c
/* comp.c -- compose a message
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/charstring.h"
#include "sbr/fmt_new.h"
#include "whatnowproc.h"
#include "sbr/m_name.h"
#include "sbr/m_gmprot.h"
#include "sbr/getarguments.h"
#include "sbr/read_switch_multiword.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_save.h"
#include "sbr/showfile.h"
#include "sbr/smatch.h"
#include "sbr/refile.h"
#include "sbr/cpydata.h"
#include "sbr/m_draft.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/fdcompare.h"
#include "sbr/folder_read.h"
#include "sbr/context_save.h"
#include "sbr/context_find.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "sbr/utils.h"
#include "sbr/fmt_compile.h"
#include "sbr/fmt_scan.h"
#include "sbr/done.h"
#include "sbr/m_maildir.h"
#include <fcntl.h>
#include "sbr/globals.h"

#define COMP_SWITCHES \
    X("draftfolder +folder", 0, DFOLDSW) \
    X("draftmessage msg", 0, DMSGSW) \
    X("nodraftfolder", 0, NDFLDSW) \
    X("editor editor", 0, EDITRSW) \
    X("noedit", 0, NEDITSW) \
    X("file file", 0, FILESW) \
    X("form formfile", 0, FORMSW) \
    X("use", 0, USESW) \
    X("nouse", 0, NUSESW) \
    X("whatnowproc program", 0, WHATSW) \
    X("nowhatnowproc", 0, NWHATSW) \
    X("build", 5, BILDSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("to address", 0, TOSW) \
    X("cc address", 0, CCSW) \
    X("from address", 0, FROMSW) \
    X("fcc mailbox", 0, FCCSW) \
    X("width columns", 0, WIDTHSW) \
    X("subject text", 0, SUBJECTSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(COMP);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(COMP, switches);
#undef X

#define DISPO_SWITCHES \
    X("quit", 0, NOSW) \
    X("replace", 0, YESW) \
    X("use", 0, USELSW) \
    X("list", 0, LISTDSW) \
    X("refile +folder", 0, REFILSW) \
    X("new", 0, NEWSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(DISPO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(DISPO, aqrunl);
#undef X

static struct swit aqrul[] = {
    { "quit", 0, NOSW },
    { "replace", 0, YESW },
    { "use", 0, USELSW },
    { "list", 0, LISTDSW },
    { "refile", 0, REFILSW },
    { NULL, 0, 0 }
};

int
main (int argc, char **argv)
{
    int use = NOUSE;
    bool nedit = false;
    bool nwhat = false;
    bool build = false;
    int i, in = NOTOK, isdf = 0, out, dat[5], format_len = 0;
    int outputlinelen = OUTPUTLINELEN;
    char *cp, *cwd, *maildir, *dfolder = NULL;
    char *ed = NULL, *file = NULL, *form = NULL;
    char *folder = NULL, *msg = NULL, buf[BUFSIZ];
    char *to = NULL, *from = NULL, *cc = NULL, *fcc = NULL, *dp;
    char *subject = NULL;
    char drft[BUFSIZ], **argp, **arguments;
    struct msgs *mp = NULL;
    struct format *fmt;
    struct stat st;

    if (nmh_init(argv[0], true, true)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			die("missing argument to %s", argp[-2]);
		    nedit = false;
		    continue;
		case NEDITSW: 
		    nedit = true;
		    continue;

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			die("missing argument to %s", argp[-2]);
		    nwhat = false;
		    continue;

		case BILDSW:
		    build = true;
		    /* FALLTHRU */
		case NWHATSW: 
		    nwhat = true;
		    continue;

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case USESW: 
		    use++;
		    continue;
		case NUSESW: 
		    use = NOUSE;
		    continue;

		case FILESW:	/* compatibility */
		    if (file)
			die("only one file at a time!");
		    if (!(file = *argp++) || *file == '-')
			die("missing argument to %s", argp[-2]);
		    isdf = NOTOK;
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			die("only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW: 
		    if (file)
			die("only one draft message at a time!");
		    if (!(file = *argp++) || *file == '-')
			die("missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case TOSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    to = addlist(to, cp);
		    continue;

		case CCSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    cc = addlist(cc, cp);
		    continue;

		case FROMSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    from = addlist(from, cp);
		    continue;

		case FCCSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dp = NULL;
		    if (*cp == '@')
			cp = dp = path(cp + 1, TSUBCWF);
		    fcc = addlist(fcc, cp);
                    free(dp);
		    continue;

		case WIDTHSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((outputlinelen = atoi(cp)) < 10)
			die("impossible width %d", outputlinelen);
		    continue;

		case SUBJECTSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    subject = cp;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    if (msg)
		die("only one message at a time!");
            msg = cp;
	}
    }

    cwd = mh_xstrdup(pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Check if we are using a draft folder
     * and have specified a message in it.
     */
    if ((dfolder || context_find ("Draft-Folder")) && !folder && msg && !file) {
	file = msg;
	msg = NULL;
    }
    if (form && (folder || msg))
	    die("can't mix forms and folders/msgs");

    cp = NULL;

    if (folder || msg) {
	/*
	 * Use a message as the "form" for the new message.
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == -1)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder, 1)))
	    die("unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    die("no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */
	seq_save (mp);

	if (mp->numsel > 1)
	    die("only one message at a time!");

	if ((in = open (form = mh_xstrdup(m_name (mp->lowsel)), O_RDONLY)) == -1)
	    adios (form, "unable to open message");
    } else {
	struct comp *cptr;

	if (! form)
	    form = components;

        cp = new_fs(form, NULL, NULL);
	format_len = strlen(cp);
	fmt_compile(cp, &fmt, 1);

	/*
	 * Set up any components that were fed to us on the command line
	 */

	if (from) {
	    cptr = fmt_findcomp("from");
	    if (cptr)
		cptr->c_text = from;
	}
	if (to) {
	    cptr = fmt_findcomp("to");
	    if (cptr)
		cptr->c_text = to;
	}
	if (cc) {
	    cptr = fmt_findcomp("cc");
	    if (cptr)
		cptr->c_text = cc;
	}
	if (fcc) {
	    cptr = fmt_findcomp("fcc");
	    if (cptr)
		cptr->c_text = fcc;
	}
	if (subject) {
	    cptr = fmt_findcomp("subject");
	    if (cptr)
		cptr->c_text = subject;
	}
    }

try_it_again:
    strncpy (drft, build ? m_maildir ("draft")
			: m_draft (dfolder, file, use, &isdf), sizeof(drft) - 1);

    /*
     * Check if we have an existing draft
     */
    if (!build && (out = open (drft, O_RDONLY)) != -1) {
	i = fdcompare (in, out);
	close (out);

	/*
	 * If we have given -use flag, or if the
	 * draft is just the same as the components
	 * file, then no need to ask any questions.
	 */
	if (use || i)
	    goto edit_it;

	if (stat (drft, &st) == -1)
	    adios (drft, "unable to stat");
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = read_switch_multiword ("\nDisposition? ",
						isdf ? aqrunl : aqrul)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrunl : aqrul)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    file = NULL;
		    use = NOUSE;
		    goto try_it_again;
		case YESW: 
		    break;
		case USELSW:
		    use++;
		    goto edit_it;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    inform("say what?");
		    break;
	    }
	}
    } else {
	if (use)
	    adios (drft, "unable to open");
    }

    if ((out = creat (drft, m_gmprot ())) == -1)
	adios (drft, "unable to create");
    if (cp) {
	charstring_t scanl;

	i = format_len + 1024;
	scanl = charstring_create (i + 2);
	dat[0] = 0;
	dat[1] = 0;
	dat[2] = 0;
	dat[3] = outputlinelen;
	dat[4] = 0;
	fmt_scan(fmt, scanl, i, dat, NULL);
	if (write(out, charstring_buffer (scanl),
		  charstring_bytes (scanl)) < 0) {
	    advise (drft, "write");
	}
	charstring_free(scanl);
    } else {
	cpydata (in, out, form, drft);
	close (in);
    }
    close (out);

edit_it:
    context_save ();	/* save the context file */

    if (nwhat)
	done (0);
    what_now (ed, nedit, use, drft, NULL, 0, NULL, NULL, 0, cwd, 0);
    done (1);
    return 1;
}
